carma-messenger-core/cpp_message/src/MobilityRequest_Message.cpp

```diff
@@ -0,0 +1,538 @@
+        cav_msgs::MobilityRequest output;
+
+        //decode results - stored in binary array
+        asn_dec_rval_t rval;
+        MessageFrame_t* message=nullptr;
+
+        //copy from vector to array
+        size_t len=binary_array.size();
+
+        uint8_t buf[len];
+        for(size_t i=0;i<len;i++){
+            buf[i]=binary_array[i];
+        }
+        
+        //use asn1c lib to decode
+        rval=uper_decode(0, &asn_DEF_MessageFrame, (void **) &message,buf, len,0,0);
+        if(rval.code==RC_OK){
+            Mobility_Header Header_constant;
+            std::string sender_id, recipient_id, sender_bsm_id, plan_id;
+            uint64_t timestamp;
+            //get sender id
+            size_t str_len=message->value.choice.TestMessage00.header.hostStaticId.size;
+            if(str_len<=Header_constant.STATIC_ID_MAX_LENGTH && str_len!=0)
+            {
+                for(size_t i=0;i<str_len;i++){
+                    sender_id +=message->value.choice.TestMessage00.header.hostStaticId.buf[i];
+                }
+            }
+            else sender_id=Header_constant.STRING_DEFAULT;
+
+            header.sender_id=sender_id;
+
+            //get recepient id
+            str_len=message->value.choice.TestMessage00.header.targetStaticId.size;
+            if(str_len<=Header_constant.STATIC_ID_MAX_LENGTH && str_len!=0)
+            {
+                for(size_t i=0;i<str_len;i++){
+                    recipient_id +=message->value.choice.TestMessage00.header.targetStaticId.buf[i];
+                }
+            }
+            else recipient_id=Header_constant.STRING_DEFAULT;
+
+            header.recipient_id=recipient_id;
+            
+            //get bsm id
+            str_len=message->value.choice.TestMessage00.header.hostBSMId.size;
+            if(str_len==Header_constant.BSM_ID_LENGTH)
+            {
+                for(size_t i=0;i<str_len;i++){
+                    sender_bsm_id +=message->value.choice.TestMessage00.header.hostBSMId.buf[i];
+                }
+            }
+            else sender_bsm_id=Header_constant.BSM_ID_DEFAULT;
+            
+            header.sender_bsm_id=sender_bsm_id;
+
+            //get plan id
+            str_len=message->value.choice.TestMessage00.header.planId.size;
+            if(str_len==Header_constant.GUID_LENGTH)
+            {
+                for(size_t i=0;i<str_len;i++){
+                    plan_id +=message->value.choice.TestMessage00.header.planId.buf[i];
+                }
+            }
+            else plan_id=Header_constant.GUID_DEFAULT;
+
+            header.plan_id=plan_id;
+            
+            //recover uint64_t timestamp from string
+            str_len=message->value.choice.TestMessage00.header.timestamp.size;
+            timestamp=0;
+            for(size_t i=0;i<str_len;i++){
+                timestamp*=10;
+                timestamp+=int(message->value.choice.TestMessage00.header.timestamp.buf[i])-'0';
```

Comment:
I think I get what's going on here, but it's probably better to use `std::stoi` here.

Reply:
I did it this way because buf  is not exactly a string, it's a uint8_t array